src/asl.ts

```diff
@@ -715,19 +733,57 @@ export class ASL {
               ResultPath: tempArrayPath,
             })!;
 
-        const initializerName = isIdentifier(stmt.initializer)
-          ? stmt.initializer.name
-          : isVariableDecl(stmt.initializer) &&
-            isIdentifier(stmt.initializer.name)
-          ? stmt.initializer.name.name
-          : undefined;
+        const initializer: ASLGraph.SubState | ASLGraph.NodeState = (() => {
+          /**ForInStmt
+           * Assign the value to $.0__[variableName].
+           * Assign the index to the variable decl. If the variable decl is an identifier, it may be carried beyond the ForIn.
+           */
+          if (isForInStmt(stmt)) {
+            const initializerName = isIdentifier(stmt.initializer)
+              ? stmt.initializer.name
+              : isVariableDecl(stmt.initializer) &&
+                isIdentifier(stmt.initializer.name)
+              ? stmt.initializer.name.name
+              : undefined;
+
+            if (initializerName === undefined) {
+              throw new SynthError(
+                ErrorCodes.Unexpected_Error,
+                "The left-hand side of a 'for...in' statement cannot be a destructuring pattern."
```

Comment:
Is this because we don't yet support it for the $$.Map.Index.Value logic?

Reply:
That is the error that typescript throws, playing with javascript, it does actually support destruction, but it will be an edge case as the left side is always a string.

![image](https://user-images.githubusercontent.com/289213/181807557-0575444d-30e4-4112-91b7-f82b26ba5a9e.png)

Reply:
why is `length` 1 there?

Reply:
length is the length of the name of the index

indices are "0", "1", "2", so length is the string length of each of the numeric strings.

`"0".length`